lib/ecto_watch/options/watcher_options.ex

```diff
@@ -0,0 +1,216 @@
+    Generic representation of an app schema.  Contains important details about a postgres table,
+    whether it's create from an Ecto schema module or from a map.
+    """
+    defstruct [
+      :pg_schema_name,
+      :table_name,
+      :primary_key,
+      :columns,
+      :association_columns,
+      :label
+    ]
+
+    def new(schema_mod) when is_atom(schema_mod) do
+      pg_schema_name =
+        case schema_mod.__schema__(:prefix) do
+          nil -> "public"
+          prefix -> prefix
+        end
+
+      table_name = "#{schema_mod.__schema__(:source)}"
+      [primary_key] = schema_mod.__schema__(:primary_key)
+
+      fields = schema_mod.__schema__(:fields)
+
+      association_columns =
+        schema_mod.__schema__(:associations)
+        |> Enum.map(&schema_mod.__schema__(:association, &1))
+        |> Enum.map(& &1.owner_key)
+
+      %__MODULE__{
+        pg_schema_name: pg_schema_name,
+        table_name: table_name,
+        primary_key: primary_key,
+        columns: fields,
+        association_columns: association_columns,
+        label: schema_mod
+      }
+    end
+
+    def new(%__MODULE__{}) do
+      raise "There is a bug!  SchemaDefinition struct was passed to new/1"
+    end
+
+    def new(opts) when is_map(opts) do
+      pg_schema_name = opts[:pg_schema_name] || "public"
+
+      %__MODULE__{
+        pg_schema_name: to_string(pg_schema_name),
+        table_name: to_string(opts.table_name),
+        primary_key: opts.primary_key,
+        columns: opts.columns,
+        association_columns: opts[:association_columns] || [],
+        label: "#{pg_schema_name}|#{opts.table_name}"
+      }
+    end
+  end
+
+  def validate({schema_definition, update_type}) do
+    validate({schema_definition, update_type, []})
+  end
+
+  def validate({schema_definition, update_type, opts}) do
+    with {:ok, schema_definition} <- validate_schema_definition(schema_definition, opts[:label]),
+         {:ok, update_type} <- validate_update_type(update_type),
+         {:ok, opts} <- validate_opts(opts, schema_definition, update_type) do
+      {:ok, {schema_definition, update_type, opts}}
+    end
+  end
+
+  def validate(other) do
+    {:error,
+     "should be either `{schema_definition, update_type}` or `{schema_definition, update_type, opts}`.  Got: #{inspect(other)}"}
+  end
+
+  def validate_schema_definition(schema_mod, _label_opt) when is_atom(schema_mod) do
+    if EctoWatch.Helpers.ecto_schema_mod?(schema_mod) do
+      {:ok, schema_mod}
+    else
+      {:error, "Expected atom to be an Ecto schema module. Got: #{inspect(schema_mod)}"}
+    end
+  end
+
+  def validate_schema_definition(opts, label_opt) when is_map(opts) do
+    schema = [
+      pg_schema_name: [
+        type: {:or, ~w[string atom]a},
+        required: false,
+        default: :public
+      ],
+      table_name: [
+        type: {:or, ~w[string atom]a},
+        required: true
+      ],
+      primary_key: [
+        type: :atom,
```

Comment:
Ah, so here it checks that the `primary_key` is an atom. That makes sense -- but the documentation is incorrect, I guess?